Processstore - Search component #46

AlexanderLichtenegger opened the pull request:
resolves #46

added:
- Menu
  - Entry "Process Store"
- Components/Modules
  - ProcessStoreSearchComponent (uses test data at the moment)
  - ProcessStoreSearchModule
- Routings
  - "processstore-search" -> "ProcessStoreSearchComponent"
- Models (Angular)
  - ProcessStore (based on ProcessStoreDTO from active feature branch)
- Services (Angular)
  - ProcessStoreService (preparation, not in use at the moment)
- Pipes
  - SimpleSearchPipe
- NPM Dependencies
  - angular-star-rating
    - css-star-rating
- Static Assets
  - "star-rating.icons.svg"

AlexanderLichtenegger committed:
added ProcessStoreSearchComponent

amarbajric commented on GUI-dev/src/app/processstore.service.ts:
```diff
@@ -0,0 +1,21 @@
+export class ProcessStoreService {
+
+    
+  restApi = window.location.protocol + "//" + window.location.hostname + ":10000/api";
```
Why are you using the `window` object here?
Please do not use something like this as we have a central service which is providing all endpoints and the associated urls.

amarbajric commented on GUI-dev/src/models/models.ts:
```diff
@@ -5,6 +5,19 @@ export class Process {
+  processCreatedAt: Date;
+  processVersion: number;
+  processPrice: number;
+  // missing in current ProcessStoreDTO in backend:
```
We are not 100% sure if we will ever show any image of the actual process.
However, it was initially a nice idea to implement something like this.
I would rather suggest to remove it when doing the final adaptations.

`processAverageRating` is a good idea, but it is also not clear how this will be calculated in the backend.
From my point of view however, it will most likely be provided via a separate endpoint which gives you the average rating of all ratings.

amarbajric commented:
In general your search component looks very promising to me.
However, I have to mention some minor aspects:
- Are the initially shown "cards", where one card displays a process, the processes with the best rating or just randomly shown?
- Which filters have you planned to implement? Furthermore, why haven't you already implemented a filter option like searching for the creators firstName or userName of the process?
- How is the width and especially height calculated for every single "card"? is the hight dependent on the shown image? I suggest to show equally sized cards as we might need to discard the idea with the image because we simply wont provide images from the backend.

Other than that, I like the layout and the currently implemented functionality. Like mentioned before, it looks very promising.

**ToDo:**
Please answer my questions mentioned above.
After that, we will determine if we need some adaptations to the PR or if the PR can be accepted (initially of course) as long as the services in the backend are not implemented

amarbajric reviewed (commented):
See comments in code and general comment.

tortmann reviewed (commented):
see other comment - reply required

tortmann commented:
I definitely agree with @amarbajric that it looks promising.

**Please explain your logic behind the current implementation as was also stated by @amarbajric comment.** 

- Are the initial cards listed alphabetically or randomly etc.?
- Which filter options are planned ?
- Which sorting options are planned ?

>equally sized cards

definitely a good idea even _with_ a potential image  (which would then have to be resized)

>average rating 

should be calculated in the backend and provided via an endpoint or added to an existing one that provides data to the search component

> window object

please shortly explain why you used this approach and change it to use the respective service of the project's microservice architecture. Thank you

AlexanderLichtenegger committed:
Merge branch 'dev' into processstore-search-component-#46

AlexanderLichtenegger committed:
added filtering for ProcessStoreSearchComponent

AlexanderLichtenegger commented:
Answers to the questions:
•	Cards: The cards are shown in the order as they come from the backend.
•	Filters: Three filters were added. “Price Below, Rating Above and Created By”.
•	Width/Height: The width is rendered dynamically. The height of the cards next to each other always have the equal size.
•	Images: The images were removed.
•	Sorting: Is not implemented yet. “Alphabetically, Rating High/Low and Price High/Low” could be used for the sorting.
•	Average Rating: Should come from the endpoint.

ToDos done:
•	Cards next to each other always have equal size
•	Filters added: "Price Below", "Rating Above", "Created By"
•	Images removed
•	Removed ProcessStoreService and used GatewayProvider instead

ToDos:
•	Average Rating received from Endpoint.
•	Input Validation for filters (filterInput is a string --> http://localhost:10000/api/store/processes?filterType=ratingAboveOrEquals&filterInput=abcde --> is possible)
•	Example Requests:
o	http://localhost:10000/api/store/processes?filterType=priceBelowOrEquals&filterInput=12
o	http://localhost:10000/api/store/processes?filterType=ratingAboveOrEquals&filterInput=3.5
o	http://localhost:10000/api/store/processes?filterType=createdBy&filterInput=tester2

AlexanderLichtenegger commented:
[processstore_testdata.sql.txt](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/5ac4a22c28a2d6001126c85c/125516c66e82c728ace21e0d46f88695208790ba9cfdb051a162e4913a5223723d05ed386189a164fd52690e050a1de853524d41ada88b2da5b46d70d9095febd66d5f75933b40a8ebd31aad1da862da11d1e9e1d30695aa2f2ff58a87453f1b07108efa5e33a0994cc951e406.txt)
SQL-Dummy-File --> remove .txt from file

stfnltnr committed:
added endpoints to process-store, for detailed information check API documentation and pull requests

Changes to be committed:
        modified:   ../Gateway/src/main/java/at/fhjoanneum/ippr/gateway/api/controller/ProcessStoreGatewayController.java
        modified:   ../Gateway/src/main/java/at/fhjoanneum/ippr/gateway/api/controller/user/HttpHeaderUser.java
	modified:   ../Gateway/src/main/java/at/fhjoanneum/ippr/gateway/api/services/impl/ProcessStoreCallerImpl.java
	modified:   src/main/java/at/fhjoanneum/ippr/processstore/controller/ProcessStoreController.java
	modified:   src/main/java/at/fhjoanneum/ippr/processstore/persistence/entities/ProcessStoreBuilder.java
	modified:   src/main/java/at/fhjoanneum/ippr/processstore/persistence/entities/ProcessStoreObjectImpl.java
	modified:   src/main/java/at/fhjoanneum/ippr/processstore/repositories/ProcessStore.java
	modified:   src/main/java/at/fhjoanneum/ippr/processstore/services/ProcessStoreService.java
	modified:   src/main/java/at/fhjoanneum/ippr/processstore/services/ProcessStoreServiceImpl.java
	modified:   ../SpringCommons/src/main/java/at/fhjoanneum/ippr/commons/dto/processstore/ProcessStoreDTO.java

Commit:
Merge branch 'dev' into processstore-search-component-#46

stfnltnr committed:
gradle.bat and gradle-wrapper.properties for W10 startup

Commit:
Merge branch 'store-service-#30' into processstore-search-component-#46

Commit:
fixed linting errors and removed mock data

Commit:
lint

amarbajric commented:
**Review**
- Looks very promising and in general the implementation is, at least from the visual perspective, good
- I do not agree with the filters and they need refactoring
  - Filtering should also be done at client side and not server side as this scenario allows for a perfect client side filtering without the need of numerous requests for every new application of a filter

**ToDo**
- Refactor and rework the filters

For now it is **APPROVED** BUT it needs some more work.
